### DISCUSSION TITLE
Validate a file type

### DISCUSSION THREAD

**Discussion by jdgamble555** [open-circle/valibot#165] [Q&A] [answered]
How would I validate a file type?

Zod has a way of doing custom types like so - https://github.com/colinhacks/zod/issues/387#issuecomment-1191390673

Is this possible in Valibot?

J

**Comment by fabian-hiller** [comment]
In Valibot you can use [`mimeType`](https://github.com/fabian-hiller/valibot/blob/main/library/src/validations/mimeType/mimeType.ts) for this.

**Reply by fabian-hiller** [reply]
And if you want to know more about pipelines, here is the guide: https://valibot.dev/guides/pipelines/

**Reply by jdgamble555** [reply]
This seems confusing.  I tried this, but it does not work. The `image` variable throws an error:
```ts
const userSchema = object({
    display_name: string([minLength(2)]),
    email: string([email()]),
    image: mimeType(['image/*'])
});
```

J

**Reply by fabian-hiller** [reply]
Wildcards are currently not supported.

**Reply by jdgamble555** [reply]
![jpg image](https://github.com/fabian-hiller/valibot/assets/8333090/f4455dfb-b12f-4370-a832-08d8487ab5fa)

I still get the same error with jpeg.

**Reply by fabian-hiller** [reply]
Please write `string([mimeType([...])])`

**Reply by jdgamble555** [reply]
![type problem](https://github.com/fabian-hiller/valibot/assets/8333090/8869c056-f011-4743-afdf-44cfc6fdbe08)
 Same issue... (A file is not really a string type though...)

**Reply by fabian-hiller** [reply]
Sorry, my bad. Please try it out with `blob(...)` or `instance(File, ...)`.

**Reply by jdgamble555** [reply]
Ok perfect!  What about the file size?

**Reply by fabian-hiller** [reply, marked as answer]
Use `mimeType` for the file type and `maxSize` for the maximum file size. You can find all other validation functions in our docs.

**Reply by jdgamble555** [reply]
Thanks! Docs just aren't clear on any complex use cases.

**Reply by fabian-hiller** [reply]
That is right. I plan to improve this.

**Comment by AndonMitev** [comment]
Hello @fabian-hiller, sorry for bothering but how exactly to handle files?

Tried with following schema and `"valibot": "^0.20.1"`

```
const NFTMetadataSchema = object({
  name: string([minLength(1, 'Please enter NFT name.')]),
  description: string([minLength(1, 'Please enter NFT description.')]),
  image: blob([
    mimeType([
      'image/jpg, image/jpeg, image/png, image/svg, image/gif, image/webp',
    ]),
  ]),
});
```

but hitting tons of errors in console when apply this kind of validation to `image` key

**Reply by AndonMitev** [reply]
PS with `image: instance(File)` i do get: `ReferenceError: File is not defined` however on the clientside

**Reply by AndonMitev** [reply]
for some reason i think when this piece of code is executed:

```
const NFTMetadataSchema = object({
  name: string([minLength(1, 'Please enter NFT name.')]),
  description: string([minLength(1, 'Please enter NFT description.')]),
  backgroundColor: string([minLength(1, 'Please select color')]),
  weapon: string([minLength(1, 'Please select weapon')]),
  image: instance(File),
});
```

window is not there yet or smth like this because if i try with:

`image: instance(window.File),` i do get 

`ReferenceError: window is not defined`

but the component it self is client just window is not initiated yet

**Reply by fabian-hiller** [reply]
`File` is not everywhere supported. Also, `File` is converted to `Blob` when transferred to the server. That's why we use `Blob`. Please try the following code:

```ts
const ImageSchema = blob([
  mimeType([
    'image/jpg',
    'image/jpeg',
    'image/png',
    'image/svg',
    'image/gif',
    'image/webp',
  ]),
]);
```

**Reply by jdgamble555** [reply]
I do this:
```ts
instance(File, [mimeType(['image/jpeg']), maxSize(0)])
```
J

**Reply by fabian-hiller** [reply]
Yes, that should work too 👍🏼
